UPGRADE-GUIDE.md

```diff
@@ -0,0 +1,465 @@
+
+Each of these events will emit a `context` object which is a unique object that
+can be used to identify a transaction over the course of its lifecycle. For
+example:
+
+```typescript
+interface StepEvent {
+  account: {
+    nonce: bigint;
+    balance: bigint;
+    stateRoot: Buffer;
+    codeHash: Buffer;
+  };
+  address: Buffer;
+  codeAddress: Buffer;
+  depth: number;
+  gasLeft: bigint;
+  gasRefund: bigint;
+  memory: Buffer;
+  memoryWordCount: bigint;
+  opcode: {
+    name: string;
+    fee: number;
+  };
+  pc: number;
+  returnStack: Buffer[];
+  stack: Buffer[];
+}
+const contexts = new Map();
+provider.on("ganache:vm:tx:before", (event: { context: {} }) => {
+  contexts.set(event.context, []);
+});
+provider.on("ganache:vm:tx:step", (event: StepEvent) => {
+  contexts.get(event.context).push(event.data);
+});
+provider.on("ganache:vm:tx:after", (event: { context: {} }) => {
+  doAThingWithThisTransactionsSteps(contexts.get(event.context));
+  contexts.delete(event.context);
+});
+```
+
+The reason this `context` is necessary is that Ganache may run multiple
+transactions simultaneously, so `"ganache:vm:tx:step"` events from different
+transactions could be intermingled.
+
+The above events will be emitted for `eth_call`, `*sendTransaction`,
+`debug_traceTransaction`, and `debug_storageRangeAt`.
+
+Currently, we do not await the event listener's return value, however, we'll
+likely enable this in the future.
+
+#### Old databases from previous versions are not compatible with v7.0.0
+
+Ganache's old database format is incompatible with this version. We've decided
+to hold off on building migration tools for this. If you will need a migration
+tool (you use the `db` flag or the `db_path` option and are unable to recreate 
+your initial DB state) please
+[open an issue](https://github.com/trufflesuite/ganache/issues/new?milestone=7.0.0)
+to let us know.
+
+#### Non-consecutive transaction nonces no longer throw an error
+
+We now support the `pendingTransactions` event and will soon support actual
+`pending` blocks.
+
+Previously, if you sent a transaction with a nonce that did not match the
+account's transaction count that transaction would be immediately rejected. In
+v7 that transaction will be placed in the node's transaction queue.
+
+You can replace these queued transactions the same way you'd replace the
+transaction on Mainnet or tests, by sending another transaction with the same
+nonce but a higher gas price.
+
+Currently the eviction mechanism is not tunable, but we plan on exposing options
+to change the behavior in the near future.
+
+Note: currently, the number of queued transactions does not have an upper bound
+and you can continue adding new transactions until your process runs out of
+memory and crashes. We consider this a memory leak and a bug. Expect this
+unbounded behavior to change in a patch-level release in the future.
+
+Note 2: if you use the persisted DB option: we have never stored unexecuted
```

Comment:
```suggestion
NOTE: if you use the persisted DB option: we have never stored unexecuted
```
We've had multiple notes in a row without them being numbered elsewhere, so I removed. Decisions, decisions, all of them wrong.